### PR TITLE
Add SFT validation eval with val_data

### PR DESCRIPTION
**Pull request by philippnormann**
## Summary

- Add optional `val_data` and `eval` config blocks to SFT.
- Run periodic validation inside SFT training and log `val/loss` and `val/num_batches`.
- Add config validation that requires `eval` and `val_data` to be set together.
- Add unit tests for config validation behavior.

## Why

Train loss alone is not enough for checkpoint selection and overfitting detection.

## Before

- No native periodic validation signal in SFT runs.

## After

- SFT can emit validation metrics at configurable intervals during training.

## Evidence

- Reverse-text run showing periodic validation logging behavior.

| train/loss | val/loss |
|---|---|
| train-loss | val-loss |
- Config used:

`sft_fullft_rtext_split_200.toml`

```toml
max_steps = 200

[ckpt]
interval = 20

[model]
name = "PrimeIntellect/Qwen3-0.6B"

[data]
name = "willcb/R1-reverse-wikipedia-paragraphs-v1-1000"
splits = ["train[:90%]"]
seq_len = 4096
batch_size = 32
shuffle = true
seed = 42

[val_data]
name = "willcb/R1-reverse-wikipedia-paragraphs-v1-1000"
splits = ["train[90%:]"]
seq_len = 4096
batch_size = 32
shuffle = false
seed = 42

[eval]
interval = 10
num_batches = 4

[optim]
lr = 2e-5
```

## Validation

- `uv run pytest tests/unit/train/sft/test_sft_eval_config.py -q`
- Unit tests cover: `eval` without `val_data` (invalid), `val_data` without `eval` (invalid), and `eval + val_data` (valid).
- 200-step reverse-text run emits `val/loss` every 10 steps as configured.

## Scope

- This PR covers periodic SFT validation evaluation and config validation.

---

> [!NOTE]
> **Medium Risk**
> Touches core SFT training-loop and dataset-loading paths; while changes are straightforward, they can impact training correctness/performance and distributed metric aggregation if edge cases are missed.
> 
> **Overview**
> Adds optional SFT validation via new `SFTValConfig` (`sft.val`) that loads a separate validation dataset and runs full-pass evaluation on a configurable interval (and optionally at step 0), logging `val/loss`.
> 
> Refactors SFT data loading by extracting `load_sft_dataset()` (expensive HF I/O) and extending `setup_dataset()` to accept a preloaded raw dataset plus `max_epochs`, enabling validation to reuse preloaded data.
> 
> Restructures the SFT training loop to centralize loss/forward-backward into helpers and to compute aggregated loss/NaN counts consistently across distributed ranks, with minor logging tweaks (e.g., only emitting `max_vio` when present).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d80bb23e4879830d7c007ff971305133ffd8269c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: tests/unit/train/sft/test_sft_eval_config.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+def test_sft_val_data_requires_cp_compatible_micro_batch_size():
+    with pytest.raises(ValueError, match="Validation micro batch size must be 1 when CP is enabled"):
+        SFTConfig(
+            model=ModelConfig(cp=2),
+            val=SFTValConfig(
+                interval=10,
+                data=SFTDataConfig(micro_batch_size=2),
+            ),
+        )
```
Comment (samsja) [inline tests/unit/train/sft/test_sft_eval_config.py]: lets remove this tests

File: src/prime_rl/configs/sft.py [hunk 2]
```diff
@@ -106,6 +106,12 @@ def validate_subsets_and_splits(self):
         return self
 
 
+class SFTValConfig(BaseConfig):
+    interval: Annotated[int, Field(ge=1, description="Run validation every N training steps.")] = 50
+    eval_on_start: Annotated[bool, Field(description="Run a validation pass before the training loop starts.")] = False
```
Comment (samsja) [inline src/prime_rl/configs/sft.py]: I don't think we want this

Comment (philippnormann) [reply src/prime_rl/configs/sft.py]: fair 👍 added it as an optional smoke test (mirrors HF Trainer), but i’ll remove it and push an update.

File: src/prime_rl/trainer/sft/train.py [hunk 3]
```diff
@@ -180,14 +185,110 @@ def train(config: SFTConfig):
...
+        if mean_loss != mean_loss:
+            logger.warning(f"Validation at step {step} had no valid tokens")
+        else:
+            logger.success(f"Validation | Step {step} | Loss: {mean_loss:.4f}")
+        monitor.log({"val/loss": mean_loss, "step": step}, step=step)
+        if was_training:
+            model.train()
+
+    if config.val is not None and config.val.eval_on_start:
+        run_validation(progress.step)
```
Comment (samsja) [inline src/prime_rl/trainer/sft/train.py]: we don't need a function here

Comment (samsja) [inline src/prime_rl/trainer/sft/train.py]: lets not do this, lets rather edit the if statement below to also include step 0 as valid if eval_on_start is set

**Review by samsja — APPROVED** [review]

**Comment by hallerite** [comment]
@philippnormann thank you for the PR! Sorry for taking so long to get it into a mergeable state – it was trickier than we expected, but I think it's in a good state now.

**Comment by philippnormann** [comment]
No worries, glad it made it in! Appreciate you and @samsja taking the time to get it into shape.

Also have #1849 open for SFT LoRA support if you get a chance to look at it.
